File: Hskl/combi.hs
import Data.List (sortBy)
import Data.Ord (comparing)

type Op = (String, Double -> Double)

ops :: [Op]
ops =
  [ ("A", \x -> x / 2.0)
  , ("B", \x -> x - 900)
  , ("C", \x -> x + 2000)
  , ("D", \x -> x * 5)
  , ("E", \x -> x + 500)
  ]

target :: Double
target = 3000

maxDepth :: Int
maxDepth = 6

dfs :: Double -> [String] -> Int -> [[String]]
dfs x path depth
  | depth > maxDepth = []
  | x == target = [path]
  | x < -10000 || x > 10000 = []
  | otherwise =
      concat [ dfs (f x) (path ++ [name]) (depth + 1)
             | (name, f) <- ops
             ]

results :: [[String]]
results = dfs 0 [] 0

-- sort by length, then lexicographically
sortedResults :: [[String]]
sortedResults =
  sortBy cmp results
  where
    cmp a b =
      compare (length a) (length b)
      <> compare (concat a) (concat b)

main :: IO ()
main = mapM_ (putStrLn . concat) sortedResults

File: Hskl/combi.js
const ops = [
    { name: "A", fn: x => x / 2.0 },
    { name: "B", fn: x => x - 900 },
    { name: "C", fn: x => x + 2000 },
    { name: "D", fn: x => x * 5 },
    { name: "E", fn: x => x + 500 },
  ];
  
  const TARGET = 3000;
  const MAX_DEPTH = 6;
  
  function dfs(x, path, depth) {
    let res = [];
  
    if (depth > MAX_DEPTH) return res;
  
    if (x === TARGET) {
      res.push(path);
      return res;
    }
  
    if (x < -10000 || x > 10000) return res;
  
    for (const op of ops) {
      const sub = dfs(op.fn(x), [...path, op.name], depth + 1);
      res = res.concat(sub);
    }
  
    return res;
  }

  const results = dfs(0, [], 0);

  results.sort((a, b) => {
    if (a.length !== b.length) return a.length - b.length;
  
    const sa = a.join("");
    const sb = b.join("");
    return sa.localeCompare(sb);
  });

  for (const r of results) {
    console.log(r.join(""));
  }